sylvia/src/types.rs

```diff
@@ -160,6 +160,159 @@ impl<'a, C: cosmwasm_std::CustomQuery, Contract> From<&'a BoundQuerier<'a, C, Co
     }
 }
 
+/// This structure represents a collection of execution methods for
+/// a Sylvia contract or interface. An instance of this structure
+/// should be created using a [Remote] object, obtained through the
+/// [Remote::executor] method. ExecutorEmptyBuilder implements the
+/// Executor traits generated by the contract and interface macros,
+/// encompassing all sv::msg(exec) methods of the specified contracts
+/// and interfaces.
+///
+/// ```rust
+/// pub mod another_contract {
+///     use cosmwasm_std::{Response, StdResult};
+///     use sylvia::contract;
+///     use sylvia::types::{ExecCtx, InstantiateCtx};
+///     
+///     pub struct AnotherContract {}
+///     
+///     #[contract]
+///     impl AnotherContract {
+///         pub fn new() -> Self { Self {} }
+///
+///         #[sv::msg(instantiate)]
+///         fn instantiate(&self, ctx: InstantiateCtx) -> StdResult<Response> {
+///             Ok(Response::new())
+///         }
+///
+///         #[sv::msg(exec)]
+///         fn exec_method(&self, ctx: ExecCtx) -> StdResult<Response> {
+///             Ok(Response::new())
+///         }
+///     }
+/// }
+///
+/// use cosmwasm_std::{Addr, Response, StdResult};
+/// use cw_storage_plus::Item;
+/// use sylvia::contract;
+/// use sylvia::types::{ExecCtx, InstantiateCtx, Remote};
+/// use another_contract::AnotherContract;
+/// use another_contract::sv::Executor;
+///
+/// pub struct Contract<'a> {
+///    pub remote_contract: Item<Remote<'a, AnotherContract>>,
+/// }
+///
+/// #[contract]
+/// impl<'a> Contract<'a> {
+///     pub fn new() -> Self {
+///         Self {
+///             remote_contract: Item::new("remote_contract"),
+///         }
+///     }
+///
+///     #[sv::msg(instantiate)]
+///     fn instantiate(&self, ctx: InstantiateCtx, remote_addr: Addr) -> StdResult<Response> {
+///         self.remote_contract.save(
+///             ctx.deps.storage,
+///             &Remote::new(remote_addr.clone()),
+///         )?;
+///         Ok(Response::new())
+///     }
+///
+///     #[sv::msg(exec)]
+///     fn call_another_contract(&self, ctx: ExecCtx) -> StdResult<Response> {
+///         let remote = self.remote_contract.load(ctx.deps.storage)?;
+///         let exec_method = remote.executor().exec_method()?.build();
+///         Ok(Response::new().add_message(exec_method))
+///     }
+/// }
+///
+/// # fn main() {}
+/// ```
+pub struct ExecutorEmptyBuilder<'a, Contract: ?Sized> {
+    contract: &'a cosmwasm_std::Addr,
+    funds: Vec<Coin>,
+    _phantom: std::marker::PhantomData<Contract>,
+}
```

Comment:
I'm considering keeping single Executor with generic state as currently we have two types with basically the same documentation. This would also keep examples for both interface and contract near the same type.

```suggestion
pub struct BuilderReady;
pub struct BuilderOngoing;

pub struct ExecutorBuilder<'a, Contract: ?Sized, BuilderStateT> {
    contract: &'a cosmwasm_std::Addr,
    funds: Vec<Coin>,
    msg: Option<Binary>,
    _phantom: std::marker::PhantomData<Contract>,
}
```

And then

```rust
impl<'a, Contract: ?Sized> ExecutorEmptyBuilder<'a, Contract, BuilderOngoing> {
    pub fn with_funds()...
}

impl<'a, Contract: ?Sized> ExecutorMethods for ExecutorBuilder<'a, Contract, BuilderOngoing> {
    pub fn some_exec()...
}

impl<'a, Contract: ?Sized> ExecutorBuilder<'a, Contract, BuilderReady> {
    pub fn build()...
}
```

Since the message is constructed with the generated methods we are sure the `build` method won't be called with the `msg` not instantiated.

Reply:
Done. By the way `trait Executor` was also improved (removed lifetime), the builder type consists of one generic type `State`.

Reply:
Nice!